src/Extensions/Abstractions/ExecutionContext/BaseExecutionContext.cs

```diff
@@ -137,6 +143,158 @@ protected static string GetBuildVersion()
+													.serviceTypeName;
+				return manifestMetaInfo.Single(entry => entry.serviceTypeName == targetServiceType).buildVersion;
+			}
+			catch (Exception)
+			{
+				// Target service not found
+				return null;
+			}
+		}
+
+		internal static IEnumerable<(string serviceTypeName, string buildVersion)> QueryServicesTypesWithVersion(IEnumerable<string> manifestNames)
+		{
+			return manifestNames
+				.Select(manifestFilename =>
+				{
+					string? manifestFilePath = GetServiceManifestPath(manifestFilename);
+					if (manifestFilePath == null)
+					{
+						return (serviceTypeName: null, buildVersion: null);
+					}
+
+					IEnumerable<XElement> manifestDescendants = XDocument.Load(manifestFilePath).Descendants();
+					IEnumerable<XElement> serviceManifestEntries = manifestDescendants
+						.Where(entry => entry.Name.LocalName == "ServiceManifest");
+
+					if (serviceManifestEntries.Count() != 1)
+					{
+						return (serviceTypeName: null, buildVersion: null);
+					}
+
+					string? buildVersion = serviceManifestEntries.Single().Attribute("Version")?.Value;
+
+					IEnumerable<string?> serviceTypeNames = manifestDescendants
+						.Where(entry => entry.Name.LocalName is "StatelessServiceType" or "StatefulServiceType")
+						.Select(entry => entry.Attribute("ServiceTypeName")?.Value);
+
+					if (serviceTypeNames.Count() != 1)
+					{
+						return (serviceTypeName: null, buildVersion: null);
+					}
+
+					return (serviceTypeName: serviceTypeNames.Single(), buildVersion);
+				})
+				.Where(entry => entry.serviceTypeName != null && entry.buildVersion != null)
+				.Select(entry => (serviceTypeName: entry.serviceTypeName!, buildVersion: entry.buildVersion!));
+		}
+
+		internal static IEnumerable<(string serviceName, string serviceTypeName)> QueryServicesNamesWithTypes(IEnumerable<XElement> descendants)
+		{
+			return descendants
+					.Where(entry => entry.Name.LocalName == "Service")
+					.Select(entry =>
+					{
+						string? serviceName = entry.Attribute("Name")?.Value;
+						IEnumerable<XElement> serviceDescendants = entry
+							.Descendants().Where(entry => entry.Name.LocalName is "StatelessService" or "StatefulService");
+
+						if (serviceDescendants.Count() != 1)
+						{
+							return (null, null);
+						}
+						XElement head = serviceDescendants.Single();
+						return (serviceName, serviceTypeName: head.Attribute("ServiceTypeName")?.Value);
+					})
+					.Where(entry => entry.serviceName != null && entry.serviceTypeName != null)
+					.Select(entry => (serviceName: entry.serviceName!, serviceTypeName: entry.serviceTypeName!)); // we know at this point that there are not nulls.
+		}
+
+		internal static string? GetApplicationManifestPath()
+		{
+			Regex regex = new(@"(ApplicationManifest\..*\.xml)$");
+			string[] files = Directory
+				.EnumerateFiles(@"C:\\SfDevCluster\", "*.*", SearchOption.AllDirectories)
```

Comment:
This part looks risky, I'll try to look into better alternatives

Reply:
We can start eventually from `C:\` but it may result in low performance because we can look up a lot of garbage.

Reply:
Is this only supposed to work on dev machines? I believe we use other drives to keep SF files in Azure clusters (D or S).